cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtilsTest.java

```diff
@@ -0,0 +1,91 @@
+
+    // Verify 3: Not low utilization and compute balance threshold lower bound
+    verifyComputingResourceUtilizationBalanceThreshold(0.3,
+                                                       resource,
+                                                       ResourceDistributionGoal.BALANCE_MARGIN,
+                                                       true,
+                                                       true,
+                                                       0.2,
+                                                       1.3,
+                                                       1.2,
+                                                       0.3 * (1 - ((1.3 * 1.2) - 1) * ResourceDistributionGoal.BALANCE_MARGIN));
+
+    // Verify 4: Not low utilization and compute balance threshold upper bound
+    verifyComputingResourceUtilizationBalanceThreshold(0.3,
+                                                       resource,
+                                                       ResourceDistributionGoal.BALANCE_MARGIN,
+                                                       false,
+                                                       true,
+                                                       0.2,
+                                                       1.3,
+                                                       1.2,
+                                                       0.3 * (1 + ((1.3 * 1.2) - 1) * ResourceDistributionGoal.BALANCE_MARGIN));
+
+  }
+
+  private void verifyComputingResourceUtilizationBalanceThreshold(double avgUtilizationPercentage,
+                                                                  Resource resource,
+                                                                  double balanceMargin,
+                                                                  boolean isLowerThreshold,
+                                                                  boolean isTriggeredByGoalViolation,
+                                                                  double lowUtilizationThreshold,
+                                                                  double resourceBalancePercentage,
+                                                                  double goalViolationDistributionThresholdMultiplier,
+                                                                  double expectedComputedBalanceThreshold) {
+
+    BalancingConstraint mockBalanceConstraint = EasyMock.mock(BalancingConstraint.class);
+    EasyMock.expect(mockBalanceConstraint.lowUtilizationThreshold(resource)).andReturn(lowUtilizationThreshold).anyTimes();
+    EasyMock.expect(mockBalanceConstraint.resourceBalancePercentage(resource)).andReturn(resourceBalancePercentage).anyTimes();
+    EasyMock.expect(mockBalanceConstraint.goalViolationDistributionThresholdMultiplier()).
+        andReturn(goalViolationDistributionThresholdMultiplier).anyTimes();
```

Comment:
Do we really expect this functions to be called `anyTimes()`? Can't we be more specific?

Reply:
I used `anyTimes()` intentionally since ensuring specific methods on this `BalancingConstraint` object gets invoked exactly N number of times is meaningless IMO.  The `BalancingConstraint` object serves as a group of constant values, as long as `computeResourceUtilizationBalanceThreshold` computes the correct value, the user of this method should not care how many times the implementation of this method tries to access constant values in the `BalancingConstraint` object.

Reply:
I don't have a strong opinion -- feel free to leave it as is.

But, we as developers (not users) may care about the number of times a function is called to improve/preserve efficiency of our implementations. For example, an implementation that calls a function within a loop and just before a loop may achieve the same outcome, but the latter implementation would be more efficient/preferable.

Reply:
I see your point. I agree that it worth verifying the number of invocations of a method if we need to measure/ensure efficiency. However, that is not necessary in this case. So, I will leave it as is.